### ISSUE TITLE
Clarify coding style guide for docstring location

### ISSUE THREAD

**Issue by karliss** [radareorg/cutter#2123]
Cutter currently contains docstring comments both in headers and next to implementation in cpp files. There are more of them in headers 273 compared to 50 in cpp files.

It might be better to have them always in header unless the function is visible only within single translation unit and doesn't have a declaration outside cpp file.

Tasks:
* Confirm if the assumption above is correct
* Test how docstring in cpp files behaves in different IDEs
* Test how doxygen reacts to docstrings in cpp files
* Update developer documentation once the best approach is clarified by steps above
* Consider cleaning up existing docstrings if one approach has significant benefit over the other

Benefits of having the docstring in header
* If docstring is in the header QtCreator displays it when autocompleting or hovering over the function call. Need to test if other tools have similar behavior. 

Benefits of having docstring next to implementation
* It is easier to keep the comment up to date when modifying the function

**Comment by ITAYC0HEN** [comment]
Google Coding Style guide differentiate between declaration and definition comments like this:
>  **Declaration comments** describe use of the function (when it is non-obvious); **comments at the definition** of a function describe operation.

From: [Google's C++ Style Guide#Function Comments](https://google.github.io/styleguide/cppguide.html#Function_Comments):

## Full explanation:

----

### Function Declarations

Almost every function declaration should have comments immediately preceding it that describe what the function does and how to use it. These comments may be omitted only if the function is simple and obvious (e.g. simple accessors for obvious properties of the class). These comments should open with descriptive verbs in the indicative mood ("Opens the file") rather than verbs in the imperative ("Open the file"). The comment describes the function; it does not tell the function what to do. In general, these comments do not describe how the function performs its task. Instead, that should be left to comments in the function definition.

**Types of things to mention in comments at the function declaration:**

- What the inputs and outputs are.
- For class member functions: whether the object remembers reference arguments beyond the duration of the method call, and whether it will free them or not.
- If the function allocates memory that the caller must free.
- Whether any of the arguments can be a null pointer.
- If there are any performance implications of how a function is used.
- If the function is re-entrant. What are its synchronization assumptions?

----
### Function Definitions

If there is anything tricky about how a function does its job, the function definition should have an explanatory comment. For example, in the definition comment you might describe any coding tricks you use, give an overview of the steps you go through, or explain why you chose to implement the function in the way you did rather than using a viable alternative. For instance, you might mention why it must acquire a lock for the first half of the function but why it is not needed for the second half.

**Note** _you should not just repeat the comments given with the function declaration, in the .h file or wherever. It's okay to recapitulate briefly what the function does, but the focus of the comments should be on how it does it._

**Comment by karliss** [comment]
Tool behaviour:
* QtCreator - shows the docstring  only from declaration
* Clion - docstring can be either place
* Vscode - docstring can be in either place, requires configuring cmake and compiling for help to start working
* Doxygen - docstring can be in either place

The recommendation from Google Coding Style guide seems to match with my initial suggestion of keeping docstring in next to declaration in header. Docstrings describe what the function does, how to use it and such comments belong to declaration.

**Comment by ITAYC0HEN** [comment]
I agree.
@thestr4ng3r @xarkes if you have nothing against deciding to use docstrings in the header files (declarations), then I think we can se it as an action-item

**Comment by thestr4ng3r** [comment]
Documentation should be in the header. What is an action-item?

**Comment by ITAYC0HEN** [comment]
https://en.wikipedia.org/wiki/Action_item

To remove all the remaining docstrings we have in CPP files to Headers
> There are more of them in headers 273 compared to 50 in cpp files.

**Comment by xarkes** [comment]
Yep I agree, although there are few disadvantages in my opinion I think the docstrings belong to the header files.
